flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/resources/ClusterResourceManager.java

```diff
@@ -0,0 +1,282 @@
+ *
+ * <p>We currently track CPU and memory. Ephemeral storage is missing because there is no easy way
+ * to get node statics on free storage.
+ */
+public class ClusterResourceManager implements ResourceCheck {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ClusterResourceManager.class);
+
+    /** ConfigMap name of the Kubernetes Cluster Autoscaler. */
+    static final String CLUSTER_AUTOSCALER_CONFIG_MAP = "cluster-autoscaler-status";
+
+    /** EKS specific node group information. Code still works without this label. */
+    static final String LABEL_NODE_GROUP = "eks.amazonaws.com/nodegroup";
+
+    private final Duration refreshInterval;
+    private final KubernetesClient kubernetesClient;
+
+    @VisibleForTesting ClusterResourceView clusterResourceView;
+
+    public ClusterResourceManager(Duration refreshInterval, KubernetesClient kubernetesClient) {
+        this.refreshInterval = refreshInterval;
+        this.kubernetesClient = kubernetesClient;
+    }
+
+    @Override
+    public synchronized boolean trySchedule(
+            int currentInstances,
+            int newInstances,
+            double cpuPerInstance,
+            double memoryPerInstance) {
+
+        if (refreshInterval.isNegative()) {
+            // Feature disabled
+            return true;
+        }
+
+        if (shouldRefreshView(clusterResourceView, refreshInterval)) {
+            try {
+                clusterResourceView = createResourceView(kubernetesClient);
+            } catch (KubernetesClientException e) {
+                if (e.getCode() == 403) {
+                    LOG.warn(
+                            "No permission to retrieve node resource usage. Resource check disabled.");
+                    return true;
+                }
+                throw e;
+            }
+        }
+
+        if (newInstances <= currentInstances) {
+            LOG.debug("Fewer or same amount of resources used after scaling.");
+            return true;
+        }
+
+        return trySchedule(
+                clusterResourceView,
+                currentInstances,
+                newInstances,
+                cpuPerInstance,
+                memoryPerInstance);
+    }
+
+    /**
+     * Simple check whether the new resource requirements can be scheduled. Note: This is not a
+     * full-blown scheduler. We may return false-negatives, i.e. we may indicate scheduling is not
+     * possible when it actually is. This is still better than false positives where we suffer from
+     * downtime due to non-schedulable pods.
+     */
+    private static boolean trySchedule(
+            ClusterResourceView resourceView,
+            int currentInstances,
+            int newInstances,
+            double cpuPerInstance,
+            double memoryPerInstance) {
+
+        resourceView.cancelPending();
+
+        for (int i = 0; i < currentInstances; i++) {
+            resourceView.release(cpuPerInstance, memoryPerInstance);
```

Comment:
If we only check for the increment in instances we wouldn't need the whole release logic right?

Reply:
We do need the release logic because otherwise we would never be able to remove resources from the internal view for scale downs. That means other pipelines might not get scheduled because the accounting is incorrect.

Reply:
True, right now this assumes a single operator per cluster.

Are you suggesting we should drop the precise bookkeeping and accept that we will yield false-negatives (block scaling) until we refresh our resource view? I guess we could do that.

Reply:
I'm a bit torn on this. With multiple operators and even without that when other compute resources also use the cluster the bookkeeping breaks down and produces false positives and false negatives anyways.

So it may be better to opt for a simpler approach and accept these limitations. What do you think?

Reply:
I agree, this is tricky. If we removed the `release` function and used a diff instead of absolute number of instances, that would only be marginally simpler. While it is true that false-negatives and false-positives are possible, in a single operator scenario, we would accept many false-negatives which we could prevent. 

I understand your argument about multiple operators, but in its current state, even the simpler version without releasing resources, wouldn't work correctly for this scenario. The only proper solution would be to make the operator aware of the quotas for the namespace it manages, or share the resource allocations via some communication channel (e.g. config map).

Reply:
Another safety measurement I had thought about, is to switch to a different mode when we are near max capacity. In this case, we would potentially go to the API for every scaling request. Of course with a rate limit.